Спринт 12.

WayOfTheDev opened the pull request:
Добавлена функциональность по лайкам и выход из профиля.

WayOfTheDev committed:
Спринт 12.

Добавлена функциональность по лайкам и выход из профиля.

WayOfTheDev committed:
Спринт 12. Финальная проверка

Helvetii951 commented on Image Feed/AppDelegate.swift:
```diff
@@ -1,35 +1,36 @@
+        _ application: UIApplication,
+        didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
+    ) -> Bool {
+        ProgressHUD.animationType = .activityIndicator
+        ProgressHUD.colorHUD = .black
+        ProgressHUD.colorAnimation = .lightGray
```
Лучше выносить подобные настройки в отдельную функцию. Иначе [application(_:didFinishLaunchingWithOptions:)](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622921-application) может стать громоздкой.

Helvetii951 commented on Image Feed/Auth/AuthViewController.swift:
```diff
@@ -1,76 +1,90 @@
     
-    // MARK: - Lifecycle Methods
+    // MARK: - UI Elements
+    private lazy var loginButton: UIButton = {
+        let button = UIButton()
+        button.translatesAutoresizingMaskIntoConstraints = false
+        button.layer.cornerRadius = 16
+        button.clipsToBounds = true
+        button.setTitle("Войти", for: .normal)
+        button.titleLabel?.font = .boldSystemFont(ofSize: 18)
+        button.setTitleColor(.ypBlack, for: .normal)
+        button.backgroundColor = .ypWhite
+        button.addTarget(self, action: #selector(didTapLoginButton), for: .touchUpInside)
+        return button
+    }()
```
Молодец, что используешь ленивые свойства. Это - очень хорошая практика.

Helvetii951 commented on Image Feed/Feed/ImagesListService.swift:
```diff
@@ -0,0 +1,148 @@
+        isFetching = true
+        let pageToFetch = currentPage
+        
+        guard let baseURL = Constants.defaultBaseURL else {
+            isFetching = false
+            print("ImagesListService.fetchPhotosNextPage: NetworkError.invalidBaseURL")
+            return
+        }
+        
+        let url = baseURL.appendingPathComponent("photos")
+        var urlComponents = URLComponents(url: url, resolvingAgainstBaseURL: false)
+        urlComponents?.queryItems = [
+            URLQueryItem(name: "page", value: "\(pageToFetch)"),
+            URLQueryItem(name: "per_page", value: "\(perPage)")
+        ]
+        
+        guard let finalURL = urlComponents?.url else {
+            isFetching = false
+            print("ImagesListService.fetchPhotosNextPage: NetworkError.invalidURLComponents - page: \(pageToFetch), perPage: \(perPage)")
+            return
+        }
+        
+        var request = URLRequest(url: finalURL)
+        request.httpMethod = "GET"
+        
+        if let token = OAuth2TokenStorage.shared.token {
+            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
+        }
```
Рекомендую вынести формирование `request` в отдельную функцию. Так код будет лучше структурирован, легче будет ориетироваться.

Helvetii951 commented on Image Feed/SetUp/UIColor+Extensions.swift:
```diff
@@ -0,0 +1,38 @@
+import UIKit
+
+extension UIColor {
+    static let ypBlack: UIColor = {
+        guard let color = UIColor(named: "YP Black") else {
+            fatalError("Цвет 'ypBlack' не найден в Assets.xcassets")
+        }
+        return color
+    }()
+    
+    static let ypWhite: UIColor = {
+        guard let color = UIColor(named: "YP White") else {
+            fatalError("Цвет 'ypWhite' не найден в Assets.xcassets")
+        }
+        return color
+    }()
```
`fatalError` лучше никогда не использовать. Может привести к очень неприятным падениям в будущем. Рекомендую использовать [assertionFailure](https://developer.apple.com/documentation/swift/assertionfaxrilure(_:file:line:)). 

А еще будет лучше вернуть какой-нибудь цвет по-умолчанию, если кастомный не инициализировался. Например, `.white`, `.black` и т.д.

WayOfTheDev committed:
Спринт 12. Финальное задание

1. Устранил проблему с созданием ISO8601DateFormatter
2. Обработал опциональность свойства createdAt
3. Устранил проблему с отображением неправильно открывавшегося списка фотографий
4. Вынес настройки ProgressHUD в отдельную функцию внутри AppDelegate
5. Добавил приватную функцию createRequest
6. Заменил fatalError на assertionFailure и добавил цвета по умолчанию
